tools/c7n_openstack/c7n_openstack/resources/server.py

```diff
@@ -231,3 +230,72 @@ def find_object_by_property(collection, k, v):
         return None
     assert len(result) == 1
     return result[0]
+
+
+@Server.filter_registry.register('security-group')
+class SecurityGroupFilter(ListItemFilter):
+    """Filters Servers based on attached security groups attributes
+
+    :example:
+
+    Finds servers with a security group that is stateful
+    .. code-block:: yaml
+
+            policies:
+              - name: server-securitygroup-stateful
+                resource: openstack.server
+                filters:
+                  - type: security-group
+                    attrs:
+                      - type: value
+                        key: stateful
+                        value: true
+
+    :example:
+
+    Finds servers with a security group that has inbound rules which
+    include 0.0.0.0/0 and any port
+
+    .. code-block:: yaml
+
+            policies:
+              - name: server-securitygroup-open-to-internet
+                resource: openstack.server
+                filters:
+                  - type: security-group
+                    key: security_group_rules
+                    attrs:
+                      - type: value
+                        key: direction
+                        value: ingress
+                      - type: value
+                        key: !port_range_min && !port_range_max
+                        value: true
+                      - type: value
+                        key: remote_ip_prefix
+                        value: '0.0.0.0/0'
+    """
+    schema = type_schema(
+        'security-group',
+        key= {'type': 'string'},
+        attrs={'$ref': '#/definitions/filters_common/list_item_attrs'}
+    )
+
+    annotate_items = True
+
+    def get_item_values(self, resource):
+        result = []
+        client = local_session(self.manager.session_factory).client()
+        attached_sgs = (client.compute.fetch_server_security_groups(resource.get('id'))
+                        .security_groups)
+        for sg_details in attached_sgs:
+            security_group = client.network.find_security_group(sg_details.get('id'),
```

Comment:
this would benefit from an sg_map[id] = sg as a population of resources should be mapping to a smaller population of security groups, and we shouldn't need to a given security group multiple times.